File: src/main/java/com/github/t1/yaml/dump/Canonicalizer.java
package com.github.t1.yaml.dump;

import com.github.t1.yaml.model.Document;
import com.github.t1.yaml.model.Mapping;
import com.github.t1.yaml.model.Mapping.Entry;
import com.github.t1.yaml.model.Scalar;
import com.github.t1.yaml.model.Scalar.Line;
import com.github.t1.yaml.model.Stream;
import com.github.t1.yaml.model.Visitor;
import lombok.RequiredArgsConstructor;

import java.util.List;

import static com.github.t1.yaml.model.Directive.YAML_VERSION;
import static java.util.stream.Collectors.joining;

@RequiredArgsConstructor
public class Canonicalizer implements Visitor {
    @Override public void visit(Stream stream) {
        stream.documents().removeIf(Document::isEmpty);
    }

    @Override public void visit(Document document) {
        if (document.directives().stream().noneMatch(YAML_VERSION::matchName))
            document.directives().add(YAML_VERSION);
        document.prefixComments().clear();
        if (document.node() == null)
            document.node(new Scalar());
        document.suffixComment(null);
    }

    @Override public void visit(Scalar scalar) {
        List<Line> lines = scalar.lines();
        if (!lines.isEmpty()) {
            String singleLine = lines.stream().map(Line::text).collect(joining(" "));
            singleLine = escape(scalar, singleLine);
            lines.clear();
            lines.add(new Line().text(singleLine));
        }
        scalar.doubleQuoted();
    }

    private String escape(Scalar scalar, String text) {
        switch (scalar.style()) {
            case DOUBLE_QUOTED:
                return text;
            case SINGLE_QUOTED:
                return text.replace("''", "'");
            case PLAIN:
                return text.replace("\"", "\\\"");
        }
        throw new RuntimeException("unreachable code");
    }

    @Override public void leave(Scalar scalar) {
        scalar.tag((scalar.isEmpty()) ? "!!null" : "!!str");
    }

    @Override public void enterMappingEntry(Mapping mapping, Entry entry) {
        entry.hasMarkedKey(true);
        // entry.hasNlAfterKey(true);
    }

    @Override public void leave(Stream stream) {
        stream.fixDocumentEndMarkers();
        if (stream.hasDocuments())
            stream.lastDocument().hasDocumentEndMarker(false);
    }
}

File: src/main/java/com/github/t1/yaml/dump/Canonicalizer.kt
package com.github.t1.yaml.dump

import com.github.t1.yaml.model.Directive.Companion.YAML_VERSION
import com.github.t1.yaml.model.Document
import com.github.t1.yaml.model.Mapping
import com.github.t1.yaml.model.Mapping.Entry
import com.github.t1.yaml.model.Scalar
import com.github.t1.yaml.model.Scalar.Line
import com.github.t1.yaml.model.Scalar.Style.DOUBLE_QUOTED
import com.github.t1.yaml.model.Scalar.Style.PLAIN
import com.github.t1.yaml.model.Scalar.Style.SINGLE_QUOTED
import com.github.t1.yaml.model.Stream
import com.github.t1.yaml.model.Visitor
import java.util.stream.Collectors.joining

class Canonicalizer : Visitor {
    override fun visit(stream: Stream) {
        stream.documents.removeIf { it.isEmpty }
    }

    override fun visit(document: Document) {
        if (document.directives.stream().noneMatch(YAML_VERSION::matchName))
            document.directives.add(YAML_VERSION)
        document.prefixComments.clear()
        if (document.node == null)
            document.node(Scalar())
        document.suffixComment = null
    }

    override fun visit(scalar: Scalar) {
        val lines = scalar.lines
        if (!lines.isEmpty()) {
            var singleLine = lines.stream().map { it.text }.collect(joining(" "))
            singleLine = escape(scalar, singleLine)
            lines.clear()
            lines.add(Line().text(singleLine))
        }
        scalar.doubleQuoted()
    }

    private fun escape(scalar: Scalar, text: String): String =
        when (scalar.style) {
            DOUBLE_QUOTED -> text
            SINGLE_QUOTED -> text.replace("''", "'")
            PLAIN -> text.replace("\"", "\\\"")
        }

    override fun leave(scalar: Scalar) {
        scalar.tag = if (scalar.isEmpty) "!!null" else "!!str"
    }

    override fun enterMappingEntry(mapping: Mapping, entry: Entry) {
        entry.hasMarkedKey = true
        // entry.setHasNlAfterKey(true);
    }

    override fun leave(stream: Stream) {
        stream.fixDocumentEndMarkers()
        if (stream.hasDocuments())
            stream.lastDocument().hasDocumentEndMarker = false
    }
}
